### PR TITLE
Send SSR errors from Next.js app

### PR DESCRIPTION
**Pull request by amoore108**
Overall changes
======
- Sends `SERVER_SIDE_REQUEST_FAILED` logs when the initial server render fails
- This should mimic the behaviour of the Express app when data fetching works, but the actual server render fails
- Moves the `SERVER_SIDE_RENDER_REQUEST_RECEIVED` log to the `_document.tsx` entry point as well, as this is a more accurate reflection of the initial server side request

Code changes
======
- Adds SSR failure logs to the initial server render in the Next.js app
- Moves `SERVER_SIDE_RENDER_REQUEST_RECEIVED` out of `getServerSideProps` and into the `getInitialProps` of the `_document.tsx` file, which is the first entry point into the application
- Adds function to remove sensitive headers from the logs
- Adds function to derive the page type to add to the logs

Testing
======
1. Pull down this branch locally
2. Run the Next.js app and force the render to fail. You can do this by adding in some deliberately broken logic, for example in the `LivePageLayout.tsx` file add the following:
```
  const myobj = { some: 'val' };

  // @ts-expect-error - test
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const myval = myobj.some[0].blocks.model;
```
This will cause the app to crash as its trying to access a property on an object that doesn't exist
3. Check the terminal logs and confirm that the `500` error is logged. It should contain something like: `"pageType":"live","status":500,"url":"/burmese/live/cx8jgp1zxylt?renderer_env=live"},"metadata":{"event":"ssr_request_failed","file":"pages/_document.js","label":"NEXTJS"},"timestamp":"2024-04-16 10:44:36.115"}`

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: ws-nextjs-app/pages/_document.page.tsx [hunk 1]
```diff
@@ -24,11 +34,30 @@ type DocProps = {
...
+
+    if (ctx.res?.statusCode === INTERNAL_SERVER_ERROR) {
+      logger.error(SERVER_SIDE_REQUEST_FAILED, {
+        status: INTERNAL_SERVER_ERROR,
+        message: ctx.res?.statusMessage,
+        url: path,
+        headers: removeSensitiveHeaders(ctx.req?.headers),
+        pageType: derivePageType(path),
+      });
+    }
```
Comment (karinathomasbbc) [inline ws-nextjs-app/pages/_document.page.tsx]: Should we calculate the headers & pageType once, and then pass it into the logger?

 ```suggestion
    const headers = removeSensitiveHeaders(ctx.req?.headers);
    const pageType = derivePageType(path);
 
    if (ctx.res?.statusCode === OK) {
      logger.debug(SERVER_SIDE_RENDER_REQUEST_RECEIVED, {
        url: path,
        headers,
        pageType,
      });
    }

    if (ctx.res?.statusCode === INTERNAL_SERVER_ERROR) {
      logger.error(SERVER_SIDE_REQUEST_FAILED, {
        status: INTERNAL_SERVER_ERROR,
        message: ctx.res?.statusMessage,
        url: path,
        headers,
        pageType,
      });
    }
```

**Review by karinathomasbbc — APPROVED** [review]

File: ws-nextjs-app/pages/_document.page.tsx (hunk 1 above)
Comment (amoore108) [reply ws-nextjs-app/pages/_document.page.tsx]: Yes! Thanks.

File: ws-nextjs-app/pages/_document.page.tsx [hunk 2]
```diff
@@ -24,11 +71,14 @@ type DocProps = {
...
   static async getInitialProps(ctx: DocumentContext) {
     const initialProps = await Document.getInitialProps(ctx);
-    const isApp = isAppPath(ctx.asPath || '');
+    const url = ctx.asPath || '';
+    const isApp = isAppPath(url);
 
     // Read env variables from the server and expose them to the client
     const clientSideEnvVariables = getProcessEnvAppVariables();
 
+    handleServerLogging(ctx);
```
Comment (HarveyPeachey) [inline ws-nextjs-app/pages/_document.page.tsx:80]: Does this code get bundled in the client-side app?

Comment (amoore108) [reply ws-nextjs-app/pages/_document.page.tsx:80]: No I don't think it should. Its not returned in the `getInitialProps` object and this function should only ever run on the server.

File: src/server/utilities/customMetrics/index.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@ import onEnvironment from '../onEnvironment';
 export type Params = {
   metricName: string;
   statusCode?: number | string;
-  pageType: PageTypes;
+  pageType: PageTypes | 'Unknown';
```
Comment (andrewscfc) [inline src/server/utilities/customMetrics/index.ts:8]: What is the unknown page type needed for?

File: ws-nextjs-app/utilities/removeSensitiveHeaders/index.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    ?.map(el => el.trim());
+
+  const headersCopy = { ...headers };
+
+  sensitiveHeadersArray?.forEach(header => {
+    if (headersCopy[header]) delete headersCopy[header];
+  });
+
+  return headersCopy;
+}
```
Comment (andrewscfc) [inline ws-nextjs-app/utilities/removeSensitiveHeaders/index.ts:15]: Just wondering why a reduce couldn't be used here?

**Review by andrewscfc — APPROVED** [review]
Nice work, few minor questions

File: src/server/utilities/customMetrics/index.ts (hunk 3 above)
Comment (amoore108) [reply src/server/utilities/customMetrics/index.ts:8]: Just replicated behaviour from the `server/index.jsx` functionality: https://github.com/bbc/simorgh/blob/fe56fc73a99a88edbb50f3709a82919887e1aa88/src/server/index.jsx#L196

Its used in the `derivePageType` function as the default/fallback if no other page type can be derived from the URL.

Although tbf I don't think I've seen this pop up in logs, although haven't gone looking for it.

File: ws-nextjs-app/utilities/removeSensitiveHeaders/index.ts (hunk 4 above)
Comment (amoore108) [reply ws-nextjs-app/utilities/removeSensitiveHeaders/index.ts:15]: Good call, changed it over to filter/reduce (although I find the reduce syntax a bit hard to reason about at times 😓 )

**Pull request merged**